src/lib/hooks/useAppState.js

```diff
@@ -0,0 +1,28 @@
+import { useEffect, useState } from 'react'
+import { AppState } from 'react-native'
+
+export default settings => {
+  const { onChange, onForeground, onBackground } = settings || {}
+  const [appState, setAppState] = useState(AppState.currentState)
+
+  useEffect(() => {
+    const handleAppStateChange = nextAppState => {
+      if (nextAppState === 'active') {
+        isValidFunction(onForeground) && onForeground()
+      } else if (appState === 'active' && nextAppState.match(/inactive|background/)) {
+        isValidFunction(onBackground) && onBackground()
+      }
+      setAppState(nextAppState)
+      appState !== nextAppState && isValidFunction(onChange) && onChange(nextAppState)
+    }
+    AppState.addEventListener('change', handleAppStateChange)
+
+    return () => AppState.removeEventListener('change', handleAppStateChange)
+  }, [settings])
```

Comment:
1. This effect wont't work as you expected, as it doesn't take into account appState changes.
so, for each settings change it will be executed once, will add AppState's listener, but inside listener will access appState value only by the execution's moment. And next state changes will be compared to this appState value. As a result, onChange will be called just once (per each `settings`  change)
2. Also there are  redundant checks (like isValidFunction, regexp in state check) could be avoided

you have correct implementation of `useActiveState` from my test assignment, i've added callbacks so you could just re-use it:

```javascript
import { useState, useEffect } from 'react';
import { AppState } from 'react-native';
import { noop } from 'lodash'

const isActiveState = state => 'active' === state;

export default (settings = {}) => {
  const {
    onActivates = noop,
    onDeactivates = noop,
    onChanges = noop
  } = settings

  const [isActive, setIsActive] = useState(
    isActiveState(AppState.currentState)
  );

  useEffect(() => {
    const onAppStateChanged = state => {
      const isAppActive = isActiveState(state);

      if (isActive !== isAppActive) {
        (isActive ? onActivates : onDeactivates)()
        onChanges(isActive)
        setIsActive(isAppActive);
      }
    };

    AppState.addEventListener('change', onAppStateChanged);
    return () => AppState.removeEventListener('change', onAppStateChanged);
  }, [isActive, onActivates, onDeactivates, onChanges]);

  return isActive;
}
```

Reply:
3. Also my hook returns isActive flag you need everywhere isntead of state enum (active / inactive / background)
So you will avoid checking state === 'active' in each component using the hook ;)

Reply:
ios has 3 states so enum is required

Reply:
my hook takes it into account:
```
isActiveState = state => 'active' === state
```

other states (inactive, background) are casted to the inactive.
then we comparing casted to isActive flag states:
```
isActive !== isAppActive
```